### DISCUSSION TITLE
Feature Discussions

### DISCUSSION THREAD

**Discussion by Mohitmadhav** [Project-Easter/Flutter-UI#121] [Ideas]
Hello Contributors 💯  

Here we'll be discussing: 
- Existing feature implementations (Most important)  
- New features (Open for suggestions)
- Feature improvements

Feel free to give inputs on how can we make the app better and improve its scalability 💥

**Comment by Mohitmadhav** [comment]
https://github.com/Project-Easter/Flutter-UI/discussions/90#discussioncomment-1429809

@keyurgit45 For the Book collection you just need to store the id of the owner of the book. Rest everything will remain the same.

But we cannot be just the uid created in Firebase because the same owner can have multiple books.
Like I have 2 books - Harry Potter and The Alchemist, now both of them cannot have the same id.

So we need to look into it. @ankitoscar what other way can be proposed?

**Reply by Mohitmadhav** [reply]
> [#90 (reply in thread)](https://github.com/Project-Easter/Flutter-UI/discussions/90#discussioncomment-1429809)
> 
> @keyurgit45 For the Book collection you just need to store the id of the owner of the book. Rest everything will remain the same.
> 
> But we cannot be just the uid created in Firebase because the same owner can have multiple books. Like I have 2 books - Harry Potter and The Alchemist, now both of them cannot have the same id.
> 
> So we need to look into it. @ankitoscar what other way can be proposed?

@keyurgit45 that unique id for the book will be the doc id of Firebase for every book added by the owner. Now sounds cool i guess

**Reply by keyurgit45** [reply]
> > [#90 (reply in thread)](https://github.com/Project-Easter/Flutter-UI/discussions/90#discussioncomment-1429809)
> > @keyurgit45 For the Book collection you just need to store the id of the owner of the book. Rest everything will remain the same.
> > But we cannot be just the uid created in Firebase because the same owner can have multiple books. Like I have 2 books - Harry Potter and The Alchemist, now both of them cannot have the same id.
> > So we need to look into it. @ankitoscar what other way can be proposed?
> 
> @keyurgit45 that unique id for the book will be the doc id of Firebase for every book added by the owner. Now sounds cool i guess

Yeah..can you send the screenshot cloud firestore once? Frankly speaking, if you database services.dart there are methods already written and I am getting confused

**Reply by Mohitmadhav** [reply]
![image](https://user-images.githubusercontent.com/59333817/136041919-42883209-50ba-4dc1-9a5f-06a7a89d37dd.png)

![image](https://user-images.githubusercontent.com/59333817/136042051-2eb6f61e-e098-4d75-ab13-857424978915.png)

**Comment by Mohitmadhav** [comment]
Yes, the methods are already created, you just need to add the missing fields

**Reply by keyurgit45** [reply]
means I have to just refractor the code. Okay👍

**Reply by keyurgit45** [reply]
![image](https://user-images.githubusercontent.com/75253653/136045086-5ee64c1b-0e3e-47d1-ab45-79461eecbe17.png)

@Mohitmadhav anything to add or remove?

**Reply by Mohitmadhav** [reply]
@keyurgit45 where will you store doc id?

**Reply by keyurgit45** [reply]
![image](https://user-images.githubusercontent.com/75253653/136075769-bfaa84b8-661a-4d18-bb80-be1eeb893aa7.png)

is this supposed to be done? why do we need a doc id? @Mohitmadhav

**Reply by Mohitmadhav** [reply]
UID is the user id of the user right?
But we need a book id to recognize that a particular book is owned by a particular person. That is the doc id.

For eg -  A book id (Doc id) will tell that I own the book - The Alchemist.

**Reply by keyurgit45** [reply]
yes UID is user id... to recognize that a particular book is owned by a particular person, there is a field, **final string owner** where we can store doc id or book id.  so doc id will be user id right?

**Reply by HARSHBHUDOLIA** [reply]
Yes, every user will be a document with auto generated unique id which will be the id that will stored in book owner field

**Reply by keyurgit45** [reply]
![image](https://user-images.githubusercontent.com/75253653/136131710-cefa8635-017d-4d14-8033-ec3d6c6b35c4.png)
@HARSHBHUDOLIA In the image, you can see, Every user is a doc with their UID, not auto generated id.  So in the book owner field which auto-generated id will be storing? or you can suggest changes in the diagram which I sent above. Thank You

**Reply by HARSHBHUDOLIA** [reply]
Every doc has a auto generated ID,  the uid is same as that auto generated id. That would be be the unique identifier for any user and also value of the field for book owner

**Reply by HARSHBHUDOLIA** [reply]
The diagram you have sent is fine, ultimately the aim is that we are able to connect all the books of a user with that user,

**Reply by keyurgit45** [reply]
Got it! Thanks @HARSHBHUDOLIA

**Reply by keyurgit45** [reply]
Also If I lend the book to another user we need to store that user's id in the book info?

**Reply by HARSHBHUDOLIA** [reply]
Earlier we were planning to have a whole system for that, but not now, because we will not be able to decide wether they exchanged the book or not after they start chatting, we are also not asking if they have exchanged the book.
So we cannot dynamically updated the lent books id's in the future versions of the app we will have features like that

**Reply by HARSHBHUDOLIA** [reply]
Please take note that uid generated after a user log in via firebase and user id are two different things.

**Reply by HARSHBHUDOLIA** [reply]
Since the code is not done, please name the field userId to avoid confusion.

**Reply by keyurgit45** [reply]
So how the user ID is calculated for user? @HARSHBHUDOLIA

**Reply by HARSHBHUDOLIA** [reply]
Already discussed right, user id is same as doc auto generated id that is created by firebase for each user

**Reply by HARSHBHUDOLIA** [reply]
I am just asking you to rename uid to user id from previous screen shot that was shown

**Reply by keyurgit45** [reply]
Okay

**Reply by keyurgit45** [reply]
so what we can do when the user will register in the app firebase will generate an auto doc id, we need to save it on a local device to access later. so when the user will add a book using that user id stored on the local device we can create a collection and then store books data under owned books with collection id ISBN number of the book into it.is it ok? @Mohitmadhav @HARSHBHUDOLIA

**Reply by HARSHBHUDOLIA** [reply]
Or whenever any user is inputting a book, you can store it in books collection, every book will have a field owner, which will store the userid of the user inputting the book

**Reply by HARSHBHUDOLIA** [reply]
As you know firebase has a collection document collection document model.

So every user(collection) document can have a collection Books which consists of all the books we have.

We can in addition to that also make a separate collection to store all the books.

This way fetching can be easier.

Inputting the data in two places.

I hope you understood

**Reply by keyurgit45** [reply]
> As you know firebase has a collection document collection document model.
> 
> So every user(collection) document can have a collection Books which consists of all the books we have.
> 
> We can in addition to that also make a separate collection to store all the books.
> 
> This way fetching can be easier.
> 
> Inputting the data in two places.
> 
> I hope you understood 

That is a good idea. To make collection for users we are taking document id from firebaseauth.instance.currentuset.uid

**Reply by Mohitmadhav** [reply]
> so what we can do when the user will register in the app firebase will generate an auto doc id, we need to save it on a local device to access later. so when the user will add a book using that user id stored on the local device we can create a collection and then store books data under-owned books with collection id ISBN number of the book into it.is it ok? @Mohitmadhav @HARSHBHUDOLIA

Yep, saving the doc id will work.
Save the doc ID using flutter_secure_storage package not shared preferences.
See :lib\utils\keys_storage.dart

**Reply by keyurgit45** [reply]
> > so what we can do when the user will register in the app firebase will generate an auto doc id, we need to save it on a local device to access later. so when the user will add a book using that user id stored on the local device we can create a collection and then store books data under-owned books with collection id ISBN number of the book into it.is it ok? @Mohitmadhav @HARSHBHUDOLIA
> 
> Yep, saving the doc id will work.
> Save the doc ID using flutter_secure_storage package not shared preferences.
> See :lib\utils\keys_storage.dart

Actually now it is sorted we dont need auto generated doc id we have firebase Auth generated user ID which is unique. We can use that for document creation @Mohitmadhav

**Comment by HARSHBHUDOLIA** [comment]
The collective doesn't need to have any id.
Every time one login the id may change so its not constant also

**Comment by HARSHBHUDOLIA** [comment]
Sorry, I was confused, yes you are right. We can go by what you said

**Reply by keyurgit45** [reply]
yeah, that's what I was asking. Okay👍

**Comment by keyurgit45** [comment]
![image](https://user-images.githubusercontent.com/75253653/136185479-696bf4a7-35d1-4898-a95c-b874ad1cd795.png)
![image](https://user-images.githubusercontent.com/75253653/136185522-f42c1118-53e1-4657-8d9d-e8476cde2df8.png)
As suggested by @HARSHBHUDOLIA we have 2 collections.. each inputted book from the user will go in both collections. is it ok? @Mohitmadhav 
note: I tested it on my database.

**Reply by Mohitmadhav** [reply]
Cool @keyurgit45

**Comment by keyurgit45** [comment]
https://user-images.githubusercontent.com/75253653/136256577-debd5717-3529-42c2-9f5e-ed9200307c41.mp4

Should I make PR so can review the code? @Mohitmadhav

**Reply by Mohitmadhav** [reply]
Yes @keyurgit45 make the PR, the rest I'll tell after reviewing the code

**Comment by Mohitmadhav** [comment]
https://github.com/Project-Easter/Flutter-UI/pull/129#issuecomment-937473520

@keyurgit45 I found the cause...Have a look at:  https://stackoverflow.com/questions/7908954/google-books-api-searching-by-isbn

And note the difference between: 
1. https://www.googleapis.com/books/v1/volumes?q=9788172234980
2. https://www.googleapis.com/books/v1/volumes?q=isbn:9788172234980

We were using the first one, but we need the second one.
Your ISBN query: https://www.googleapis.com/books/v1/volumes?q=isbn:9780062416216

Both of them are  - The Alchemist

Now you can resolve this.

**Reply by Mohitmadhav** [reply]
I also enabled Books API in GCP for our project, let's see how that goes.

**Reply by keyurgit45** [reply]
what is GCP?

**Reply by keyurgit45** [reply]
> [#129 (comment)](https://github.com/Project-Easter/Flutter-UI/pull/129#issuecomment-937473520)
> 
> @keyurgit45 I found the cause...Have a look at: https://stackoverflow.com/questions/7908954/google-books-api-searching-by-isbn
> 
> And note the difference between:
> 
> 1. https://www.googleapis.com/books/v1/volumes?q=9788172234980
> 2. https://www.googleapis.com/books/v1/volumes?q=isbn:9788172234980
> 
> We were using the first one, but we need the second one. Your ISBN query: https://www.googleapis.com/books/v1/volumes?q=isbn:9780062416216
> 
> Both of them are - The Alchemist
> 
> Now you can resolve this.

Yeah also found the same. I will make the changes

**Reply by Mohitmadhav** [reply]
> what is GCP?

Google Cloud Platform

**Reply by keyurgit45** [reply]
There is one problem. when we use this query we don't get the thumbnail image.
actually there are two books the alchemist one from 2015 and one from 2005 so there are two ISBN numbers. one has thumbnail IMG and one doesn't have.

**Reply by keyurgit45** [reply]
> There is one problem. when we use this query we don't get the thumbnail image.
> actually there are two books the alchemist one from 2015 and one from 2005 so there are two ISBN numbers. one has thumbnail IMG and one doesn't have.

Can we show no image found photo if we don't found thumbnail

**Reply by Mohitmadhav** [reply]
Yes, you did that already I guess, in the makeBookforDB function

**Reply by keyurgit45** [reply]
> Yes, you did that already I guess, in the makeBookforDB function

Yes.. There is some need of modification though.. Will make it and will send PR

**Comment by Mohitmadhav** [comment]
@keyurgit45 I got an exceptional case:
https://www.googleapis.com/books/v1/volumes?q=isbn:9789388754439

**Reply by Mohitmadhav** [reply]
So I changed a bit of code in the main. You can pull from there

**Reply by keyurgit45** [reply]
ohhh a book without an author.. we need to check the null value error for each field then.

**Comment by Mohitmadhav** [comment]
@keyurgit45 the bookmark is working now, you can pull it from main.
And yes, I'll be creating some issues tonight

**Comment by keyurgit45** [comment]
@Mohitmadhav Regarding the user_preferences issue, the chips are not clickable. It is happening because of the package, so I tried changing the packages which support null safety but no one worked so can we write our own implementation for that?

**Reply by Mohitmadhav** [reply]
Yes, I was just going to create an issue.
Anyways @keyurgit45 since Hacktoberfest is over now, it's just a matter of a few glitches along with the Chat feature, then we are good to go for the Play Store release.

**Reply by Mohitmadhav** [reply]
You can create an issue for chip choice as a continuation of #130

**Reply by keyurgit45** [reply]
I have created an issue. What to do for book recommendations using genres? @Mohitmadhav

**Reply by Mohitmadhav** [reply]
Let's resolve this null safety chip choice and then generate random books according to genre if not all

**Reply by keyurgit45** [reply]
@Mohitmadhav who fixed multiple chip choice issue? . I just pulled changes from repo.

**Reply by Mohitmadhav** [reply]
No, it's not fixed yet. Is it working on your system?

**Reply by keyurgit45** [reply]
Yup , I guess someone modified the code but one thing not working after saving genres when we open again the dialog it should select all genres which are saved previously. Should I go with this code or I have created my own class and its working fine .what should I do? @Mohitmadhav

**Reply by Mohitmadhav** [reply]
Yeah, you need to modify that firstly.

**Reply by keyurgit45** [reply]
I have made changes in the existing code I haven't implemented a new class as the existing code was working with some bugs. please go through. @Mohitmadhav

**Comment by Mohitmadhav** [comment]
@keyurgit45 lets look at the explore screen and then chat.
Btw I resolved the personalize button error

**Reply by keyurgit45** [reply]
Yes, I tried a few changes but didn't work. Do you have any idea how to solve this? @Mohitmadhav

**Comment by Mohitmadhav** [comment]
Hey @keyurgit45, sorry for the late reply. I was busy with college academics and stuff. Can you take up the Firebase Chat issue after which we can release it in the play store? (I'll look into other bugs and features after 15th)

**Reply by keyurgit45** [reply]
@Mohitmadhav actually i have my semester exams from 20th so can I take it up after exams?

**Reply by Mohitmadhav** [reply]
Sure @keyurgit45

**Reply by Mohitmadhav** [reply]
@keyurgit45 when can you resume again?

**Reply by keyurgit45** [reply]
> @keyurgit45 when can you resume again?

from today itself.

**Reply by keyurgit45** [reply]
@Mohitmadhav it is completed. I have tested this feature using my firebase app but for your firebase app, it is giving me errors while accessing chats collection.

"[ERROR:flutter/lib/ui/ui_dart_state.cc(209)] Unhandled Exception: [cloud_firestore/permission-denied] The caller does not have permission to execute the specified operation."

**Reply by Mohitmadhav** [reply]
I've given you the Firebase access. Now you can check it out.

**Reply by keyurgit45** [reply]
please check now, we are using collection chat instead of chats

**Reply by Mohitmadhav** [reply]
@keyurgit45 are you receiving the chat from my side?

**Reply by keyurgit45** [reply]
@Mohitmadhav yeah. I have sent you back

**Reply by Mohitmadhav** [reply]
Yeah @keyurgit45 but now there's another problem...None of the books are showing the owners name, and the public profile route is not setup

**Reply by Mohitmadhav** [reply]
> Yeah @keyurgit45 but now there's another problem...None of the books are showing the owners name, and the public profile route is not setup

So without that we cannot send message to anyone else who owns a book

**Reply by keyurgit45** [reply]
I checked in the database. these books are missing some fields that why it is why giving error... and in explore tab books are increasing automatically after revisiting the tab again so please check that issue also
@Mohitmadhav

**Reply by Mohitmadhav** [reply]
Which part is missing?? The entries are the same as in the database file where we created the collections...Maybe it's time to again check the database design

**Reply by Mohitmadhav** [reply]
![image](https://user-images.githubusercontent.com/59333817/151539259-53367aa7-a9d9-43b2-8afc-6afde326418d.png)
RequestThrottledException ([Error]: (006) Request Throttled. Over Rate limit: up to 2 per sec. See geocode.xyz/pricing)
https://geocode.xyz/pricing

Even this not free

**Reply by keyurgit45** [reply]
> Which part is missing?? The entries are the same as in the database file where we created the collections...Maybe it's time to again check the database design

see the books which you have added have only 6 fields whereas the book I have added contains all fields. i don't understand why it is happening. @Mohitmadhav

**Reply by keyurgit45** [reply]
> ![image](https://user-images.githubusercontent.com/59333817/151539259-53367aa7-a9d9-43b2-8afc-6afde326418d.png) RequestThrottledException ([Error]: (006) Request Throttled. Over Rate limit: up to 2 per sec. See geocode.xyz/pricing) https://geocode.xyz/pricing
> 
> Even this not free

we can use https://pub.dev/packages/geocoding, it gives name, street, country, postal code, administrative area, sub-admin area, locality, sub locality.

**Reply by Mohitmadhav** [reply]
> > Which part is missing?? The entries are the same as in the database file where we created the collections...Maybe it's time to again check the database design
> 
> see the books which you have added have only 6 fields whereas the book I have added contains all fields. i don't understand why it is happening. @Mohitmadhav

@keyurgit45 I have deleted all the entries in the Cloud firestore....lemme check now

**Reply by keyurgit45** [reply]
@Mohitmadhav any updates?

**Reply by Mohitmadhav** [reply]
Ohh yes @keyurgit45, it's been a while since I have checked the project due to Exams and stuff.
It was working back then...but just pull the latest code from main and check it once.

**Reply by keyurgit45** [reply]
pub get is showing the issue : 
This app is using a deprecated version of the Android embedding.
To avoid unexpected runtime failures or future build failures, try to migrate this app to the V2 embedding.
Take a look at the docs for migrating an app: https://github.com/flutter/flutter/wiki/Upgrading-pre-1.12-Android-projects
The plugin `cloud_firestore` requires your app to be migrated to the Android embedding v2.

**Reply by Mohitmadhav** [reply]
@keyurgit45 I tried to solve this problem but I am having some issues myself with Gradle. Apart from that, we are in our last month of college. The exam, project reviews, and farewells have kept me busy. But we have decided to make a new UI/UX in the month of May/June/July that will enhance the existing interface. This time I will do it more professionally, so I request you to wait till I am ready to start working on it. If you want to we can discuss the plans with you in detail on a call or a meet link.

**Reply by keyurgit45** [reply]
Yeah sure. Take your time. In my free time, I will try to look into the error. @Mohitmadhav

**Reply by HARSHBHUDOLIA** [reply]
@keyurgit45  We plan to have it released in the play store and then work on the revamp version of the app, do you have any friends who can help us with the completion of the app. There are some bugs which we find hard to solve and the deployment part which we have little knowledge about. 80% of the app is done with some expertise rest can be done too. If possible see if you can take help of someone to get the things done.

**Reply by keyurgit45** [reply]
> @keyurgit45  We plan to have it released in the play store and then work on the revamp version of the app, do you have any friends who can help us with the completion of the app. There are some bugs which we find hard to solve and the deployment part which we have little knowledge about. 80% of the app is done with some expertise rest can be done too. If possible see if you can take help of someone to get the things done.

Is there any deadline to upload app to play store? I have told my friends from another college to look into it but not sure if they will help or not.

**Comment by HARSHBHUDOLIA** [comment]
@keyurgit45 No specific deadline but we plan to launch it in playstore by the month of June, while simulataneously @Mohitmadhav will be working on UI/UX to have a more professional feel to it.
